### PR TITLE
JDK-8298138: Shenandoah: HdrSeq asserts "sub-bucket index (512) overflow for value ( 1.00)"

### PR DESCRIPTION
**Pull request by ysramakrishna**
JBS link: https://bugs.openjdk.org/browse/JDK-8298138
- Fixed a boundary condition that was triggering an assert.
- Added a simple-minded gtest for HdrSeq, which allows one to exercise the asserting code in a debug build.
- Tested with: `CONF=slowdebug make run-test TEST="gtest:BasicShenandoahNumberSeqTest"`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298138](https://bugs.openjdk.org/browse/JDK-8298138): Shenandoah: HdrSeq asserts "sub-bucket index (512) overflow for value (    1.00)"

### Reviewers
 * [Roman Kennke](https://openjdk.org/census#rkennke) (@rkennke - **Reviewer**)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11524/head:pull/11524` \
`$ git checkout pull/11524`

Update a local copy of the PR: \
`$ git checkout pull/11524` \
`$ git pull https://git.openjdk.org/jdk pull/11524/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11524`

View PR using the GUI difftool: \
`$ git pr show -t 11524`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11524.diff">https://git.openjdk.org/jdk/pull/11524.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/gtest/gc/shenandoah/test_shenandoahNumberSeq.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2016, 2017, Oracle and/or its affiliates. All rights reserved.
```
Comment (rkennke) [inline test/hotspot/gtest/gc/shenandoah/test_shenandoahNumberSeq.cpp:2]: The copyright should be 2022.

**Review by rkennke — CHANGES_REQUESTED** [review]
Hi Ramki,
The change looks good. I have a few minor comments.

File: src/hotspot/share/gc/shenandoah/shenandoahNumberSeq.hpp [hunk 2]
```diff
@@ -55,7 +55,7 @@ class HdrSeq: public NumberSeq {
 // Binary magnitude sequence stores the power-of-two histogram.
 // It has very low memory requirements, and is thread-safe. When accuracy
 // is not needed, it is preferred over HdrSeq.
-class BinaryMagnitudeSeq {
+class BinaryMagnitudeSeq : public CHeapObj<mtInternal> {
```
[the remark this answers is not in the document]
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahNumberSeq.hpp]: Wanted a spec for allocation for correct accounting.

Changed to mtGC; thanks!

File: test/hotspot/gtest/gc/shenandoah/test_shenandoahNumberSeq.cpp (hunk 1 above)
Comment (ysramakrishna) [reply test/hotspot/gtest/gc/shenandoah/test_shenandoahNumberSeq.cpp]: Fixed; thanks for the catch!

**Comment by ysramakrishna** [comment]
@rkennke & @shipilev : could you folks please review and approve? I made the changes requested by Roman.

Thanks!
-- Ramki

**Review by shipilev — APPROVED** [review]
So the failure is: we want the bucket to cover `[a; a+n)`, but current code makes it cover `[a; a+n]`, which means the right-most value would overflow its assignment for sub-bucket? If so, the fix looks good.

**Review by rkennke — APPROVED** [review]
Looks good, thank you!

**Comment by ysramakrishna** [comment]
Thanks for the reviews, Roman and Alexey!
